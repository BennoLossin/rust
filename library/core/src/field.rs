//! Field Reflection

/// Type representing a field of a `struct`.
///
/// # Safety
///
/// Given a valid value of type `Self::Base`, there exists a valid value of type `Self::Type` at
/// byte offset `OFFSET`.
#[lang = "UnalignedField"]
#[unstable(feature = "field_projections", issue = "145383")]
pub unsafe trait UnalignedField: Sized {
    /// The type of the base where this field exists in.
    type Base: ?Sized;

    /// The type of the field.
    type Type: ?Sized;

    /// The offset of the field in bytes.
    const OFFSET: usize;
}

/// Type representing an aligned field of a `struct`.
///
/// # Safety
///
/// Given a well-aligned value of type `Self::Base`, the field at `Self::OFFSET` of type
/// `Self::Type` is well-aligned.
#[lang = "Field"]
#[unstable(feature = "field_projections", issue = "145383")]
pub unsafe trait Field: UnalignedField {}

/// Type representing a field with structural pinning information.
///
/// # Safety
///
/// `Self::Projected<'a>` either is `Pin<&'a mut Self::Type>` or `&'a mut Self::Type`. In the first
/// case the field is structurally pinned.
#[lang = "PinnableField"]
#[unstable(feature = "field_projections", issue = "145383")]
pub unsafe trait PinnableField: UnalignedField {
    /// The pin-projection of this field.
    ///
    /// If this field is structurally pinned, this is `Pin<P>` otherwise it is `P`.
    type Projected<P>;

    /// Sets the correct value for a pin projection.
    ///
    /// # Safety
    ///
    /// The supplied reference must be derived from a `Pin<&mut Self::Base>`.
    unsafe fn from_pinned_ref(r: &mut Self::Type) -> Self::Projected<&mut Self::Type>;
}

/// TODO
#[unstable(feature = "field_projections", issue = "145383")]
#[allow_internal_unstable(builtin_syntax)]
pub macro field_of($Container:ty, $($fields:expr)+ $(,)?) {
    builtin # field_of($Container, $($fields)+)
}
