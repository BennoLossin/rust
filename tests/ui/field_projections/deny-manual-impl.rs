#![allow(incomplete_features)]
#![feature(field_projections)]

use std::field::{Field, UnalignedField, field_of};

#[repr(packed)]
pub struct MyStruct(usize);

unsafe impl UnalignedField for MyStruct {
    //~^ ERROR: the `UnalignedField` trait may not be implemented manually [E0806]
    type Base = ();
    type Type = ();
    const OFFSET: usize = 0;
}

unsafe impl Field for field_of!(MyStruct, 0) {} //~ ERROR: the `Field` trait may not be implemented manually [E0806]

fn main() {}
