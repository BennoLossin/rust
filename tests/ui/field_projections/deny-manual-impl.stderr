error[E0806]: the `UnalignedField` trait may not be implemented manually
  --> $DIR/deny-manual-impl.rs:9:1
   |
LL | unsafe impl UnalignedField for MyStruct {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0806]: the `Field` trait may not be implemented manually
  --> $DIR/deny-manual-impl.rs:16:1
   |
LL | unsafe impl Field for field_of!(MyStruct, 0) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0806`.
